src/media/local-stream.ts

```diff
@@ -0,0 +1,194 @@
+
+  // The output stream can change to reflect any effects that have
+  // been added.  This member will always point to the MediaStream
+  // that this LocalStream was originally created with.
+  protected inputStream: MediaStream;
+
+  /**
+   * Create a LocalStream from the given values.
+   *
+   * @param stream - The initial output MediaStream for this Stream.
+   */
+  constructor(stream: MediaStream) {
+    super(stream);
+    this.inputStream = stream;
+  }
+
+  /**
+   * Get the track within the MediaStream with which this LocalStream was created.
+   *
+   * @returns The track within the MediaStream with which this LocalStream
+   * was created.
+   */
+  protected get inputTrack(): MediaStreamTrack {
+    return this.inputStream.getTracks()[0];
+  }
+
+  /**
+   * @inheritdoc
+   */
+  get muted(): boolean {
+    return !this.inputTrack.enabled;
+  }
+
+  /**
+   * Set the mute state of this stream.
+   *
+   * @param isMuted - True to mute, false to unmute.
+   */
+  setMuted(isMuted: boolean): void {
+    if (this.inputTrack.enabled === isMuted) {
+      this.inputTrack.enabled = !isMuted;
+      // setting `enabled` will not automatically emit MuteStateChange, so we emit it here
+      this[StreamEventNames.MuteStateChange].emit(isMuted);
+    }
+  }
+
+  /**
+   * @inheritdoc
+   */
+  getSettings(): MediaTrackSettings {
+    return this.inputTrack.getSettings();
+  }
+
+  /**
+   * Get the label of the input track on this stream.
+   *
+   * @returns The label of the track.
+   */
+  get label(): string {
+    return this.inputTrack.label;
+  }
+
+  /**
+   * Change the track of the output stream to a different track.
+   *
+   * Note: this method assumes and enforces that if both input and output streams have the same
+   * track, then they must also be the same stream.
+   *
+   * @param newTrack - The track to be used in the output stream.
+   */
+  private changeOutputTrack(newTrack: MediaStreamTrack): void {
+    if (this.outputTrack.id !== newTrack.id) {
+      // If the input track and the *old* output track are currently the same, then the streams must
+      // be the same too. We want to apply the new track to the output stream without affecting the
+      // input stream, so we separate them by setting the input stream to be its own stream.
+      if (this.inputTrack.id === this.outputTrack.id) {
+        this.inputStream = new MediaStream(this.inputStream);
+      }
+
+      this.outputStream.removeTrack(this.outputTrack);
+      this.outputStream.addTrack(newTrack);
+
+      // If the input track and the *new* output track are now the same, then we want the streams to
+      // be the same too.
+      if (this.inputTrack.id === this.outputTrack.id) {
+        this.inputStream = this.outputStream;
+      }
+
+      this[LocalStreamEventNames.OutputTrackChange].emit(newTrack);
+    }
+  }
+
+  /**
+   * @inheritdoc
+   */
+  stop(): void {
+    this.inputTrack.stop();
```

Comment:
do we need to stop also the outputTrack?

Reply:
Yeah we should. Good catch, will fix it.